apps/reaper/test/unit/reaper/data_extract/extract_step_test.exs

```diff
@@ -0,0 +1,538 @@
+            headers: %{},
+            cacheTtl: nil
+          },
+          assigns: %{
+            key: "super secret"
+          }
+        }
+      ]
+
+      assigns = ExtractStep.execute_extract_steps(dataset, steps)
+
+      assert assigns == %{key: "super secret", token: "auth_token"}
+    end
+
+    test "Can use assigns block for headers", %{bypass: bypass, dataset: dataset} do
+      Bypass.stub(bypass, "POST", "/headers", fn conn ->
+        if Enum.any?(conn.req_headers, fn header -> header == {"header", "super secret"} end) do
+          Plug.Conn.resp(conn, 200, %{sub: %{path: "auth_token"}} |> Jason.encode!())
+        else
+          Plug.Conn.resp(conn, 401, "Unauthorized")
+        end
+      end)
+
+      steps = [
+        %{
+          type: "auth",
+          context: %{
+            path: ["sub", "path"],
+            destination: "token",
+            url: "http://localhost:#{bypass.port}/headers",
+            encodeMethod: "json",
+            body: %{},
+            headers: %{Header: "{{header}}"},
+            cacheTtl: nil
+          },
+          assigns: %{
+            header: "super secret"
+          }
+        }
+      ]
+
+      assigns = ExtractStep.execute_extract_steps(dataset, steps)
+
+      assert assigns == %{header: "super secret", token: "auth_token"}
+    end
+
+    test "Can use assigns block for url", %{bypass: bypass, dataset: dataset} do
+      Bypass.stub(bypass, "POST", "/fancyurl", fn conn ->
+        Plug.Conn.resp(conn, 200, %{sub: %{path: "auth_token"}} |> Jason.encode!())
+      end)
+
+      steps = [
+        %{
+          type: "auth",
+          context: %{
+            path: ["sub", "path"],
+            destination: "token",
+            url: "http://localhost:#{bypass.port}/{{path}}",
+            encodeMethod: "json",
+            body: %{},
+            headers: %{},
+            cacheTtl: nil
+          },
+          assigns: %{
+            path: "fancyurl"
+          }
+        }
+      ]
+
+      assigns = ExtractStep.execute_extract_steps(dataset, steps)
+
+      assert assigns == %{path: "fancyurl", token: "auth_token"}
+    end
+
+    test "fails with a reasonable error message", %{bypass: bypass, dataset: dataset} do
+      Bypass.stub(bypass, "POST", "/", fn conn ->
+        Plug.Conn.resp(conn, 403, "No dice")
+      end)
+
+      steps = [
+        %{
+          type: "auth",
+          context: %{
+            path: ["sub", "path"],
+            destination: "token",
+            url: "http://localhost:#{bypass.port}",
+            encodeMethod: "json",
+            body: %{Key: "AuthToken"},
+            headers: %{},
+            cacheTtl: nil
+          },
+          assigns: %{}
+        }
+      ]
+
+      assert_raise RuntimeError, "Unable to process auth step for dataset 12345-6789.", fn ->
+        ExtractStep.execute_extract_steps(dataset, steps)
+      end
+    end
+  end
+
+  describe "execute_extract_steps/2 date" do
```

Comment:
I wouldn't mind putting the date stuff through it's paces, or at least having a test show what happens if it fails. I "feel" like it's not going to send back bad info, but we having some part of that be caught would be great. One thing we suffer from really bad with Reaper right now is the decode being the only piece that fails and then we're scratching our head about wtf is returning an HTML page. If anything, I would want to make sure this things throws when you give it bogus deltas, which could be here or in a unit test

Reply:
you found this below